src/lightning/fabric/strategies/launchers/subprocess_script.py

```diff
@@ -159,3 +170,53 @@ def _hydra_subprocess_cmd(local_rank: int) -> Tuple[Sequence[str], str]:
     # Set output_subdir null since we don't want different subprocesses trying to write to config.yaml
     command += [f"hydra.run.dir={rundir}", f"hydra.job.name=train_ddp_process_{local_rank}", "hydra.output_subdir=null"]
     return command, cwd
+
+
+def _launch_process_observer(child_processes: List[subprocess.Popen]) -> None:
+    """Launches a thread that runs along the main process and monitors the health of all processes."""
+    monitor_thread = Thread(
+        target=_ChildProcessObserver(child_processes=child_processes, main_pid=os.getpid()),
+        daemon=True,  # thread stops if the main process exits
+    )
+    monitor_thread.start()
+
+
+class _ChildProcessObserver:
+    def __init__(self, main_pid: int, child_processes: List[subprocess.Popen], sleep_period: int = 5) -> None:
+        self._main_pid = main_pid
+        self._child_processes = child_processes
+        self._sleep_period = sleep_period
+        # Note: SIGTERM is not aggressive enough to terminate processes hanging in collectives
+        self._termination_signal = signal.SIGTERM if sys.platform == "win32" else signal.SIGKILL
+        self._finished = False
+
+    def __call__(self) -> None:
+        while not self._finished:
+            time.sleep(self._sleep_period)
+            self._finished = self._run()
+
+    def _run(self) -> bool:
+        """Runs once over all child processes to check whether they are still running."""
+        for proc in self._child_processes:
+            proc.poll()
+
+        return_codes = [proc.returncode for proc in self._child_processes]
+        if all(return_code == 0 for return_code in return_codes):
+            return True
+
+        for proc in self._child_processes:
+            if proc.returncode:
+                _logger.info(
+                    f"Child process with PID {proc.pid} terminated with code {proc.returncode}."
+                    f" Forcefully terminating all other processes to avoid zombies 🧟"
+                )
+                self._terminate_all()
```

Comment:
Doesn't this partially repeat what's in https://github.com/Lightning-AI/lightning/blob/master/src/lightning/pytorch/trainer/connectors/signal_connector.py#L107-L115 already?

Reply:
It is conceptually different. This logic in the signal connector ensures that all processes are terminated when a SIGTERM arrives. This means the process is being terminated in a regular fashion by an external signal, and we are just ensuring all (child) processes get terminated completely.

Here however, we are monitoring on the side if any of the processes exit unexpectedly while others are still running. This prevents a stall of the program. 

Adding it to the multiprocessing launcher is also possible, the logic will be slightly different.

Reply:
Couldn't this also include the logic to terminate subprocesses if it's the main process that gets killed (normally by a signal)? Consolidating this logic together would be useful

Reply:
That doesn't fit well into a thread. 1) there is no need to have a thread if the main process can kill the processes that it have launched. 2) If the main process gets terminated, the thread would also be terminated and so it wouldn't be able to run its logic. I don't see a need to consolidate or abstract this.